### DISCUSSION TITLE
Anycable gRPC problem with Google Cloud Run

### DISCUSSION THREAD

**Discussion by vaarsn** [anycable/anycable#151] [Q&A] [answered]
Hi to all,
I'm experiencing an issue with Anycable gRPC and Anycable-go interconnection on Google Cloud Run service. I'm exposing the default port which is 50051 gRPC to Cloud Run. Anycable-go work without issues there but not gRPC one. When I'm trying to check the service even visiting the browser, it gives me the error: `upstream connect error or disconnect/reset before headers. reset reason: remote resetI enabled HTTP2 for that container`. I'm not using any nginx or any other proxies. I spent more than 2 days while trying to solve that issue but no success. 
Just to sum up.  I have 2 separate containers hosted with Cloud Run: `anycable-rpc` and `anycable-go`. Anycable-go works fine and it's able to connect to Anycable RPC hosted on VM but it's not working with Anycable RPC hosted on Cloud Run. I'm using the port 50051 to expose RPC  on Cloud Run which terminates it TLS then so Anycable RPC should be available via port number 443 but it's not available by fact.

Here's my **RPC-Dockerfile** which is exposed by Google on https://cloudrunhost.us.a.run.app:443:
```
FROM ruby:2.7.1-alpine
RUN apk update \
     && apk --no-cache --update add build-base git gcompat supervisor postgresql-dev
RUN rm -rf /var/cache/apk/*
ENV BUNDLE_JOBS=6 \
  BUNDLE_PATH=/bundle \
  RAILS_ENV=production \
  TZ=America/New_York 
WORKDIR /usr/src/app
ADD . .
RUN gem install bundler:2.2.27
RUN bundle install
CMD ["bundle", "exec", "anycable", "--rpc-host=0.0.0.0:50051"]
EXPOSE 50051
```
Here's my **Anycable-go envs**:
```
REDIS_URL=redis://10.97.192.3:6379
ANYCABLE_HOST=0.0.0.0 
ANYCABLE_RPC_HOST=cloudrunhost.us.a.run.app:443
```
Here's the logs produced by Cloud Run:
```
In the same time in logs I'm getting the next continuous messages:
EESTgRPC version: 1.38.0
EESThandling /anycable.RPC/Connect with #<Method: AnyCable::GRPC::Handler#connect(request, call) /bundle/ruby/2.7.0/gems/anycable-core-1.1.1/lib/anycable/grpc/handler.rb:12>
EESThandling /anycable.RPC/Command with #<Method: AnyCable::GRPC::Handler#command(request, call) /bundle/ruby/2.7.0/gems/anycable-core-1.1.1/lib/anycable/grpc/handler.rb:20>
EESThandling /anycable.RPC/Disconnect with #<Method: AnyCable::GRPC::Handler#disconnect(request, call) /bundle/ruby/2.7.0/gems/anycable-core-1.1.1/lib/anycable/grpc/handler.rb:16>
EESThandling /grpc.health.v1.Health/Check with #<Method: Grpc::Health::Checker#check(req, _call) /bundle/ruby/2.7.0/gems/grpc-1.38.0-x86_64-linux/src/ruby/pb/grpc/health/checker.rb:34>
EESThandling /grpc.health.v1.Health/Watch with #<Method: Grpc::Health::Checker(Grpc::Health::V1::Health::Service)#watch(*) /bundle/ruby/2.7.0/gems/grpc-1.38.0-x86_64-linux/src/ruby/lib/grpc/generic/service.rb:99>
EESTBroadcasting Redis channel: __anycable__
EESTRPC server is starting...
EESTRPC server is listening on 0.0.0.0:8080 (workers_num: 30)
Error
EESTGET50038 B31.7 sChrome 101 https://rpc-server-test1-kjhbtyk2aq-uc.a.run.app/health
EESTStopping... SIGTERM received
EESTstopping, will wait for all the workers to exit
EEST#<GRPC::RpcServer:0x0000557077fb4ef0>
EESTstopped, all workers are shutdown
EESTRPC server stopped
EESTStopped. Good-bye!
EESTStopping... SIGTERM received
EESTstopping, will wait for all the workers to exit

```
Please help

**Comment by palkan** [comment]
As far as I understand, you need to configure RPC server to listen on the Cloud Run specified port: https://cloud.google.com/run/docs/container-contract

It's better to rely on the $PORT env var (anycable-go uses it by default).

Something like this should help:

```dockerfile
CMD ["bundle", "exec", "anycable", "--rpc-host=0.0.0.0:$PORT"]
```

**Reply by vaarsn** [reply]
By default Google Cloud injects its port number (8080) into the docker container if it hasn't been pointed explicitly via PORT variable. That's fine. In such case if your app is listening 8080 it will be spined up by Cloud Run.
In my previous post I mentioned RPC log of the container that up and running already while its port 8080 is terminated by Google's TLS 443. But for some reason I can't connect to the RPC by using e.g. rpc.googlecloud.com:443. I'm getting the next continuous error from my Anycable-go container `sid=TJJ8oa4gVwXJzsgAhM70C WebSocket session failed: grpc connection is not ready`. RPC docker continuously shutting down. In my browser when I'm trying to open RPC server endpoint which is https://rpc.googlecloud.com:443 I'm getting the next error `upstream connect error or disconnect/reset before headers. reset reason: remote reset` Even when I've tried to use both Anycable-go and RPC on the single docker container it's not working as expected on Cloud Run but works fine locally. Here's the command I'm using to run both of them:
```
command         = bundle exec anycable --rpc-host=0.0.0.0:50051
command         = anycable-go --host=0.0.0.0 --rpc_host=0.0.0.0:50051 --redis_url=redis://myremoteredis:6379
```
It looks like RPC can't be used on Cloud Run at all. I tried with Alpine and Ubuntu backed docker images but no success.

**Comment by Envek** [comment, marked as answer]
> I'm using the port 50051 to expose RPC on Cloud Run which terminates it TLS then so Anycable RPC should be available via port number 443 but it's not available by fact.

It seems that TLS isn't enabled for communication between anycable-go and anycable-rpc.

Please either add `--rpc_enable_tls` param to anycable-go command or set `ANYCABLE_RPC_ENABLE_TLS` environment variable to `true`.

See https://docs.anycable.io/anycable-go/configuration?id=tls

**Reply by vaarsn** [reply]
It was the trick. Thanks.
